pstoreds/addr_book_gc.go

```diff
@@ -0,0 +1,299 @@
+package pstoreds
+
+import (
+	"fmt"
+	"strconv"
+	"sync/atomic"
+	"time"
+
+	ds "github.com/ipfs/go-datastore"
+	query "github.com/ipfs/go-datastore/query"
+	errors "github.com/pkg/errors"
+
+	peer "github.com/libp2p/go-libp2p-peer"
+	pb "github.com/libp2p/go-libp2p-peerstore/pb"
+
+	b32 "github.com/whyrusleeping/base32"
+)
+
+var (
+	// GC lookahead entries are stored in keys with pattern:
+	// /peers/gc/addrs/<unix timestamp of next visit>/<peer ID b32> => nil
+	gcLookaheadBase = ds.NewKey("/peers/gc/addrs")
+	// in GC routines, how many operations do we place in a batch before it's committed.
+	gcOpsPerBatch = 20
+	// queries
+	purgeQuery             = query.Query{Prefix: gcLookaheadBase.String(), KeysOnly: true}
+	populateLookaheadQuery = query.Query{Prefix: addrBookBase.String(), KeysOnly: true}
+)
+
+// cyclicBatch is similar to go-datastore autobatch, but it's driven by an actual Batch facility offered by the
+// datastore. It populates an ongoing batch with operations and automatically flushes it after N pending operations
+// have been reached. `N` is currently hardcoded to 20. An explicit `Commit()` closes this cyclic batch, erroring all
+// further operations.
+type cyclicBatch struct {
+	ds.Batch
+	ds      ds.Batching
+	pending int
+}
+
+func newCyclicBatch(ds ds.Batching) (ds.Batch, error) {
+	batch, err := ds.Batch()
+	if err != nil {
+		return nil, err
+	}
+	return &cyclicBatch{Batch: batch, ds: ds}, nil
+}
+
+func (cb *cyclicBatch) cycle() (err error) {
+	if cb.Batch == nil {
+		return errors.New("cyclic batch is closed")
+	}
+	if cb.pending < gcOpsPerBatch {
+		// we haven't reached the threshold yet.
+		return nil
+	}
+	// commit and renew the batch.
+	if err = cb.Batch.Commit(); err != nil {
+		return errors.Wrap(err, "failed while committing cyclic batch")
+	}
+	if cb.Batch, err = cb.ds.Batch(); err != nil {
+		return errors.Wrap(err, "failed while renewing cyclic batch")
+	}
+	return nil
+}
+
+func (cb *cyclicBatch) Put(key ds.Key, val []byte) error {
+	if err := cb.cycle(); err != nil {
+		return err
+	}
+	cb.pending++
+	return cb.Batch.Put(key, val)
+}
+
+func (cb *cyclicBatch) Delete(key ds.Key) error {
+	if err := cb.cycle(); err != nil {
+		return err
+	}
+	cb.pending++
+	return cb.Batch.Delete(key)
+}
+
+func (cb *cyclicBatch) Commit() error {
+	if cb.Batch == nil {
+		return errors.New("cyclic batch is closed")
+	}
+	if err := cb.Batch.Commit(); err != nil {
+		return err
+	}
+	cb.pending = 0
+	cb.Batch = nil
+	return nil
+}
+
+// purgeCycle runs a single GC cycle, operating within the lookahead window.
+//
+// It scans the lookahead region for entries that need to be visited, and performs a Clean() on them. An errors trigger
+// the removal of the GC entry, in order to prevent unactionable items from accumulating. If the error happened to be
+// temporary, the entry will be revisited in the next lookahead window.
+func (ab *dsAddrBook) purgeCycle() {
+	if atomic.LoadInt32(&ab.gcLookaheadRunning) > 0 {
```

Comment:
We shouldn't need any of these checks given that GC is being controlled from a central event loop.

Reply:
I added this guard was for robustness/defensiveness, as it prevents potentially really long purge cycles from overlapping with one another (which could also happen if the user configures a very high frequency).

Reply:
No matter *how* short the cycle is, these calls can't overlap. `purgeCycle` and `populateLookahead` are called synchronously from the same goroutine.

Reply:
That's true. However I think in the future we'll want to be able to trigger GC manually (e.g. when we start, or via a libp2p management console).